### DISCUSSION TITLE
Text_truncate in middle of word?

### DISCUSSION THREAD

**Discussion by rica-carv** [e107inc/e107#5278] [Q&A Core]
Shouldn't the `text_truncate` only truncate text after a word? Right now is truncating in the middle of words:
![image](https://github.com/e107inc/e107/assets/17130530/1c764113-5802-478e-8c0e-8b0d7bcc845a)

I'm trying to use the `text_truncate` inside the batch [news_shortcodes.php](https://github.com/e107inc/e107/blob/master/e107_core/shortcodes/batch/news_shortcodes.php) file

**Comment by Vodhin** [comment]
Text_truncate is a really "dumb" function - you give it the text, and how many characters to show before it cuts off. It could be better - but slower- if it exploded spaces to make an array of words, loop through each word (counting characters) and rep-assembling the sentence up to the last word that fit.

```
function text_truncate2($text,$max=30){
   if(strlen($text) <= $max){return $text;}
   $words = explode(" ",$text);
   if(count($words) < 2){return substr($text,0,$max).' ...';}
   $newText = array();
   $chct = 0;
   foreach($words as $word){
    if($chct >= $max){return implode(" ",$newText).' ...';}
    array_push($newText,$word);
    chct = chct +(strlen($word) +1);
    }
}
```

Personally, I use CSS to cut the text off of an element

```
.truncateMe{
   display: block;
  position: relative;
   width:100%;
   overflow-x: hidden;
   text-overflow: ellipsis;
   }
```
```

<div style="width:256px;">
   <div class="truncateMe">
    Some really long and overflowing sentence that no one would read more than the first couple of words and why are you still reading this because this is a really, really long and unneeded demonstration of the css code
   </div>
</div>
```

**Reply by rica-carv** [reply]
> Personally, I use CSS to cut the text off of an element

@Vodhin I know CSS truncate version, but the issue, my issue, is that the text continues in the html, so anyonae who want to read the full text, can do so without leaving the page or do a refresh. For me CSS obfuscation is a child play.

On the other hand, the PHP truncate is a more "adult" and "smarties" proof approach, since it leaves the code as it's seen on the screen....

Imagine you're truncating a 10000 words text....

I know these days the size of the HTML doesn't matter, but sometimes, it does matter...

**Comment by Jimmi08** [comment]
@rica-carv 
Notice year of creation.  2021!  No answer, no milestone. No discussion. 

https://github.com/e107inc/e107/issues/4480

**Reply by rica-carv** [reply]
@Jimmi08 Indeed.... 

Will you suggest a code change on the truncate function?

**Reply by Jimmi08** [reply]
No, I don't need this anymore. That code is from efiction cms, use it freely.

**Comment by Jimmi08** [comment]
My new suggestion to solve this:  - inspired by WP, they are best in this

- add element word_count_type to language XML file
- add **public** property **get_word_count_type**() to language handler   - with get and set methods
default "words",  other options:  'characters_excluding_spaces', 'characters_including_spaces' 
comment why do it this way:   
```
* translators: If your word count is based on single characters (e.g. East Asian characters),
 * enter 'characters_excluding_spaces' or 'characters_including_spaces'. Otherwise, enter 'words'.
 * Do not translate into your own language.
```

See? As English programmers  forgot about complicated European languages, we forgot about East languages too.

Then add new method to parser f.e. e_trim_words() ...  not change existing one

**Reply by rica-carv** [reply]
@Jimmi08 So it would be added a new method to parser, with barely the same code as the text_truncate method?

Right now, my pull request #5306 does both truncates (new word count and old character count truncate) with a minimal and simpler code change, and without adding a new method. It even can possibly parse east languages, just give it a try, and check.
Just my two cents...
